pkg/master/publish.go

```diff
@@ -147,11 +152,61 @@ func (m *Master) setEndpoints(serviceName string, ip net.IP, port int) error {
 			},
 		}
 	}
-	if !reflect.DeepEqual(e.Subsets, want) {
-		e.Subsets = want
-		glog.Infof("setting endpoints for master service %q to %v", serviceName, e)
+
+	// First, determine if the endpoint is in the format we expect (one
+	// subset, one port, N IP addresses).
+	formatCorrect, ipCorrect := m.checkEndpointSubsetFormat(e, ip.String(), port)
+	if !formatCorrect {
+		// Something is egregiously wrong, just re-make the endpoints record.
+		e.Subsets = []api.EndpointSubset{{
+			Addresses: []api.EndpointAddress{{IP: ip.String()}},
+			Ports:     []api.EndpointPort{{Port: port, Protocol: api.ProtocolTCP}},
+		}}
+		glog.Infof("Resetting endpoints for master service %q to %v", serviceName, e)
+		return m.endpointRegistry.UpdateEndpoints(ctx, e)
+	} else if !ipCorrect {
+		// So we *always* add our own IP address; if there are too many
+		// IP addresses, we remove the one lexicographically after our
+		// own IP address.  Given the requirements stated at the top of
+		// this function, this should cause the list of IP addresses to
+		// become eventually correct.
+		e.Subsets[0].Addresses = append(e.Subsets[0].Addresses, api.EndpointAddress{IP: ip.String()})
+		e.Subsets = endpoints.RepackSubsets(e.Subsets)
+		addrs := &e.Subsets[0].Addresses
+		for i, addr := range *addrs {
```

Comment:
It's minor, but this will make copies of every address struct.  I've gotten into the habit (as suggested by r) to do

```
for i := range *addrs {
    addr := &(*addrs)[i]
    //...
}